store/memcache.go

```diff
@@ -94,40 +99,74 @@ func (s *MemcacheStore) Set(ctx context.Context, key interface{}, value interfac
 }
 
 func (s *MemcacheStore) setTags(ctx context.Context, key interface{}, tags []string) {
+	group, ctx := errgroup.WithContext(ctx)
 	for _, tag := range tags {
-		var tagKey = fmt.Sprintf(MemcacheTagPattern, tag)
-		var cacheKeys = []string{}
-
-		if result, err := s.Get(ctx, tagKey); err == nil {
-			if bytes, ok := result.([]byte); ok {
-				cacheKeys = strings.Split(string(bytes), ",")
+		currentTag := tag
+		group.Go(func() error {
+			var tagKey = fmt.Sprintf(MemcacheTagPattern, currentTag)
+
+			var err error
+			for i := 0; i < 3; i++ {
+				if err = s.addKeyToTagValue(tagKey, key); err == nil {
+					return nil
+				}
+				// loop to retry any failure (including race conditions)
```

Comment:
I think race conditions should not be handled like this. Could you please use a mutex instead?

Reply:
@eko sorry i shouldve specified what the race condition is. its a distributed race condition, because multiple clients can update the same key at any time

previously the setTags code was using the `Set` operation to update tags. the problem with this is that if client1 issues a `Set` after client2 has read the contents of the same key via `Get`. client2 will miss the update issued by client1. this means client2 will overwrite what client1 wrote and the update is lost. 

to avoid this, synchronization is done via memcached's optimistic concurrency control mechanisms. using `CompareAndSwap` with the item returned from a previous `Get` ensures that the value has not changed since. if the item has changed, an error is returned. thus why `addKeyToTagValue` is retried

Reply:
Ok, thank you for clarifying this! So let's use this 3-times retry mechanism :)